lint/evalOrder_checker.go

```diff
@@ -0,0 +1,156 @@
+package lint
+
+//! Detects potentially unsafe dependencies on evaluation order.
+//
+// @Before:
+// return mayModifySlice(&xs), xs[0]
+//
+// @After:
+// // A)
+// v := mayModifySlice(&xs)
+// return v, xs[0]
+// // B)
+// v := xs[0]
+// return mayModifySlice(&xs), v
+
+import (
+	"go/ast"
+	"go/types"
+
+	"github.com/go-critic/go-critic/lint/internal/lintutil"
+)
+
+func init() {
+	addChecker(&evalOrderChecker{}, attrExperimental)
+}
+
+type evalOrderChecker struct {
+	checkerBase
+
+	// passive represents dependencis that are not harmful on their
+	// own, but can result in undefined result if there is
+	// equal active dependency for them.
+	passive []ast.Expr
+	// active represents depencies that can cause undefined evaluation
+	// result if there are any equal active or passive dependency for any of them.
+	active []ast.Expr
+	depSet lintutil.AstSet
+}
+
+func (c *evalOrderChecker) VisitStmt(stmt ast.Stmt) {
+	switch stmt := stmt.(type) {
+	case *ast.ReturnStmt:
+		c.checkReturn(stmt)
+	case *ast.AssignStmt:
+		// TODO(quasilyte):
+		//	https://github.com/golang/go/issues/23188
+		//	https://github.com/golang/go/issues/24448
+		//	https://github.com/golang/go/issues/23017
```

Comment:
+ https://github.com/golang/go/issues/25609 ?

Reply:
It's referenced inside test files.